### PR TITLE
Cypress: Implement SWAP mode for PSoC6 - rev2

### PR DESCRIPTION
**Pull request by BohdanKovalchuk**
Signed-off by Bohdan Kovalchuk <bohdan.kovalchuk@linaro.org>

### CODE REVIEW

**Comment by utzig** [comment]
I honestly have no idea what has been changed from the previous PR, apart from the `bootutil` changes that are not included anymore. Also by closing the previous PR and opening a new one, the previous comments stay with the old one. What about opening it only when you're fairly certain that it is OK? Or maybe adding some context of what is being done; or push by commit instead of squashing it all the time would be helpful. Is it supposed to be reviewed in the first place or should we wait for a non-draft PR?

**Comment by BohdanKovalchuk** [comment]
Thank you for responding on this quickly, Fabio.

Sorry I was not on time to add more context on this.

1. We discarded previous PR as non-acceptable by design;
2. You're right - we made SWAP mode operating on PSoC6 without changing original MCUBoot's \bootutil;
3. I assume this one can be considered as a non-draft unless you find issues in design/code.
4. Main approach here is emulating 1-byte align and write for PSoC6, so SWAP status table can be filled in correctly;

We still think there should be better overall design accepting large align values (as stated in issue #713). We plan to take a look on this early Q3 2020.

Thanks.

File: boot/cypress/BlinkyApp/main.c [hunk 1]
```diff
@@ -101,14 +101,59 @@ void test_app_init_hardware(void)
...
+    row_addr = (address/CY_FLASH_SIZEOF_ROW)*CY_FLASH_SIZEOF_ROW;
+
+    /* preserving Row */
+    memcpy((void *)row_buff, (void *)row_addr, sizeof(row_buff));
+
+    /* Modifying the target byte */
+    row_buff[address%CY_FLASH_SIZEOF_ROW] = src;
+
+    /* Programming updated row back, NO Erase */
+    rc = Cy_Flash_ProgramRow(row_addr, (const uint32_t *)row_buff);
```
[the remark this answers is not in the document]
Comment (BohdanKovalchuk) [reply boot/cypress/BlinkyApp/main.c:124]: There are two API in PSoC6's PDL driver:

_Cy_Flash_WriteRow()_

It writes an array of data to a single row of flash. This is done in three steps - pre-program, erase and then program flash row with the input data. 

_Cy_Flash_ProgramRow()_

It writes an array of data to a single row of flash.

For the given purpose we just need to set imageOK at a place were erased value sits after swap completed. Physically it OR's the data bits with flash content.

As we do not erase Row there shouldn't be loss of any data in power-failure.

File: boot/cypress/BlinkyApp/main.c [hunk 2]
```diff
@@ -101,14 +101,59 @@ void test_app_init_hardware(void)
...
-
+#ifdef SWAP_CONFIRM
+    #define USER_SWAP_IMAGE_OK_OFFS (24)
+    #define USER_SWAP_IMAGE_OK      (1)
+    uint32_t img_ok_addr;
+    int rc;
+    /* Write Image OK flag to the slot trailer, so MCUBoot-loader
+     * will not revert new image */
+    img_ok_addr = USER_APP_START+USER_SLOT_SIZE-USER_SWAP_IMAGE_OK_OFFS;
+    if(*((uint8_t *)img_ok_addr) != USER_SWAP_IMAGE_OK)
```
[the remark this answers is not in the document]
Comment (BohdanKovalchuk) [reply boot/cypress/BlinkyApp/main.c]: Fixed it.

File: boot/cypress/cy_flash_pal/cy_flash_map.c [hunk 3]
```diff
@@ -253,8 +262,89 @@ int flash_area_read(const struct flash_area *fa, uint32_t off, void *dst,
...
+                else
+                {   /* fill in buffer w erased value, so row contents will not be effected by row Program */
+                    writeBuffer[dstIndex] = CY_BOOT_INTERNAL_FLASH_ERASE_VALUE;
+                }
+                byteOffset++;
+            }
+            if(rowsNotEqual != 0u)
+            {
+                /* Program flash row */
+                rc = Cy_Flash_ProgramRow((rowId * CY_FLASH_SIZEOF_ROW), (const uint32_t *)writeBuffer);
```
[the remark this answers is not in the document]
Comment (BohdanKovalchuk) [reply boot/cypress/cy_flash_pal/cy_flash_map.c:324]: We checked the code and found out none of flash_write() is executed without complementary flash_erase().
 
So, the approach is the same - we "write back" the erased_val for non-modifiable flash contents so they are not affected - neither erased nor lost during power failure.

File: boot/cypress/BlinkyApp/main.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (d3zd3z) [reply boot/cypress/BlinkyApp/main.c:124]: I don't know about your specific device, but this is explicitly against the documentation for ProgramRow. In most devices, overwrites of existing data, if allowed at all, are only allowed a small number of times.

This also suggests that the psoc6 doesn't use any kind of ECC, which interestingly, is usually the reason for larger write sizes. So, this solution, if it actually works, would only work on psoc6.

Can you find out from hardware people if this overwriting is actually safe, despite being specifically forbidden by the documentation?

[the remark this answers is not in the document]
Comment (BohdanKovalchuk) [reply boot/cypress/BlinkyApp/main.c:124]: OK, I understood your concerns.

I'll try to get this information and comment a bit later.

**Comment by BohdanKovalchuk** [comment]
As per discussion here and w PSoC6-experts the flash_program workaround is found to be unacceptable.

Cancelling the PR.

**Pull request closed without merging**